src/init.cpp

```diff
@@ -1386,6 +1421,14 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
                     break;
                 }
 
+                if (!fReindex) {
+                    uiInterface.InitMessage(_("Rewinding blocks..."));
+                    if (!RewindBlockIndex(chainparams)) {
+                        strLoadError = _("Unable to rewind the database to a pre-upgrade state. You will need to redownload the blockchain");
```

Comment:
Everytime we are running the client we try to rewind to a pre-upgrade state? What for?

Reply:
It looks like ``RewindBlockIndex`` is supposed to do nothing if no rewind is needed, but its name (and the init message "Rewinding blocks...") does not reflect that.

Reply:
I'll see if I can think of a better name ('RewindBlockIndexIfNecessary`?) and UI message, but FWIW upstream also uses this message. I expect in most cases no one notices it, because it is immediately followed by `Verifying blocks...`.

Reply:
The main thing that confuses me is the error message.

Reply:
The main thing that confused me is the error message, it implies you are always trying to rewind till a pre-upgrade state, whereas iirc, e.g.,
if you are missing data like branch id from your last 5 blocks, but the -6 block is after upgrade but isn't missing data, you'll stop rewinding there.